### PR TITLE
Compositional crux-mir

### PR DESCRIPTION
**Pull request by spernsteiner**
This branch adds `crux-mir-comp`, a variant of `crux-mir` that supports compositional reasoning.  This is a proof-of-concept implementation, which includes the essential functionality but is missing some usability features.

For details of the design, see [`crux-mir-comp/DESIGN.md`](https://github.com/GaloisInc/saw-script/blob/sp/crux-mir-comp/crux-mir-comp/DESIGN.md).

The corresponding crucible/crux-mir changes are in GaloisInc/crucible#659.

### CODE REVIEW

File: crux-mir-comp/src/Mir/Compositional/Override.hs [hunk 1]
```diff
@@ -0,0 +1,628 @@
...
+    let postAllocDefs = filter (\(k,_v) -> not $ Map.member k preAllocMap) $
+            Map.toList $ ms ^. MS.csPostState . MS.csAllocs
+    postAllocMap <- liftM Map.fromList $ forM postAllocDefs $ \(alloc, Some allocSpec) -> do
+        ref <- newMirRefSim (allocSpec ^. maType)
+        return (alloc, Some $ MirPointer (allocSpec ^. maType) ref)
+    let allocMap = preAllocMap <> postAllocMap
+
+    -- Handle return value and post-state PointsTos
+
+    -- TODO: clobber all writable memory that's accessible in the pre state
```
Comment (atomb) [inline crux-mir-comp/src/Mir/Compositional/Override.hs:306]: How easily could this be implemented now? Or should we leave it until later?

**Review by atomb — APPROVED** [review]
This looks great to me. I left a few comments that might suggest changes, but I'll leave that up to you. There were also some aspects of the MIR-related details that I probably can't properly review due to my lack of familiarity with the underlying MIR implementation for Crucible. I appreciate how many tests you included, though, and if all of those are passing I'm quite happy to merge this.

**Comment by atomb** [comment]
Following on my test comment, I'd love to see those tests integrated into CI, but I know that'll require adding some substantial chunks to the configuration to make sure all the Rust tools are installed.

File: crux-mir-comp/src/Mir/Compositional/Override.hs (hunk 1 above)
Comment (spernsteiner) [reply crux-mir-comp/src/Mir/Compositional/Override.hs]: I actually did it already, but forgot to remove the comment.  The mechanism is a little bit indirect: for each accessible location, `add_arg` not only adds a pre-state `PointsTo` (representing a constraint on the contents of the initial memory), but also adds a post-state `PointsTo` that causes the value to be overwritten with a fresh symbolic variable.

**Comment by spernsteiner** [comment]
> Following on my test comment, I'd love to see those tests integrated into CI, but I know that'll require adding some substantial chunks to the configuration to make sure all the Rust tools are installed.

Should I try copying over the relevant parts of the GaloisInc/crucible CI config?  Or is the setup going to be complex enough that I should just leave it as a separate issue?

**Comment by atomb** [comment]
> > Following on my test comment, I'd love to see those tests integrated into CI, but I know that'll require adding some substantial chunks to the configuration to make sure all the Rust tools are installed.
> 
> Should I try copying over the relevant parts of the GaloisInc/crucible CI config? Or is the setup going to be complex enough that I should just leave it as a separate issue?

I think it's reasonable to leave that for a separate issue. We might also be able to get help from @LisannaAtGalois at some point.

**Pull request merged**